cmd/snap-discard-ns/snap-discard-ns.c

```diff
@@ -15,41 +15,190 @@
+			 "%s\\.mnt", snap_instance_name);
+	sc_must_snprintf(usr_mnt_pattern, sizeof usr_mnt_pattern,
+			 "%s\\.*\\.mnt", snap_instance_name);
+
+	DIR *ns_dir = fdopendir(ns_dir_fd);
+	if (ns_dir == NULL) {
+		die("cannot fdopendir");
+	}
+	/* ns_dir_fd is now owned by ns_dir and will not be closed. */
+
+	while (true) {
+		/* Reset errno ahead of any call to readdir to differentiate errors
+		 * from legitimate end of directory. */
+		errno = 0;
+		struct dirent *dent = readdir(ns_dir);
+		if (dent == NULL) {
+			if (errno != 0) {
+				die("cannot read next directory entry");
+			}
+			/* We've seen the whole directory. */
+			break;
+		}
+
+		/* We use dnet->d_name a lot so let's shorten it. */
+		const char *dname = dent->d_name;
+
+		/* Check the four patterns that we have against the name and set the
+		 * two should flags to decide further actions. Note that we always
+		 * unlink matching files so that is not reflected in the structure. */
+		bool should_unmount = false;
+		bool should_unlink = false;
+		struct variant {
+			const char *p;
+			bool unmount;
+		};
+		struct variant variants[4] = {
+			{.p = sys_mnt_pattern,.unmount = true},
+			{.p = usr_mnt_pattern,.unmount = true},
+			{.p = sys_fstab_pattern},
+			{.p = usr_fstab_pattern},
+		};
+		for (size_t i = 0; i < sizeof variants / sizeof *variants; ++i) {
+			struct variant *v = &variants[i];
+			debug("checking if %s matches %s", dname, v->p);
+			int match_result = fnmatch(v->p, dname, 0);
+			if (match_result < 0) {
+				die("cannot execute match against pattern %s",
+				    v->p);
+			}
+			if (match_result == 0) {
+				should_unmount |= v->unmount;
+				should_unlink = true;
+				debug("file %s matches pattern %s", dname,
+				      v->p);
+				/* One match is enough. */
+				break;
+			}
+		}
+
+		/* Stat the candidate directory entry to know what we are dealing with. */
+		struct stat file_info;
+		if (fstatat(ns_dir_fd, dname, &file_info,
+			    AT_SYMLINK_NOFOLLOW) < 0) {
+			die("cannot inspect file %s", dname);
+		}
+
+		/* We are only interested in regular files. The .mnt files, even if
+		 * bind-mounted, appear as regular files and not as symbolic links due
+		 * to the peculiarities of the Linux kernel. */
+		if (!S_ISREG(file_info.st_mode)) {
+			continue;
+		}
+
+		if (should_unmount) {
+			/* If we are asked to unmount the file double check that it is
+			 * really a preserved mount namespace since the error code from
+			 * umount2(2) is inconclusive. */
+			int path_fd = openat(ns_dir_fd, dname,
+					     O_PATH | O_CLOEXEC | O_NOFOLLOW);
+			if (path_fd < 0) {
+				die("cannot open path %s", dname);
+			}
+			struct statfs fs_info;
+			if (fstatfs(path_fd, &fs_info) < 0) {
+				die("cannot inspect file-system at %s", dname);
+			}
+			close(path_fd);
+			if (fs_info.f_type == NSFS_MAGIC
+			    || fs_info.f_type == PROC_SUPER_MAGIC) {
+				debug("unmounting %s", dname);
+				if (umount2(dname, MNT_DETACH | UMOUNT_NOFOLLOW)
+				    < 0) {
```

Comment:
This would read better if `< 0) {` was on the same line where the `if` starts, but I presume it was formatted automatically like this? Slightly odd given that the comment block above is much wider.

Reply:
This is automatically formatted, the mystery of `indent`. Perhaps given that this a near-full rewrite I could use clang-format for this file and actually switch to a sane format profile?

Reply:
That might be a good idea... Feel free to give it a try before landing and apply if it produces sensible output.